snapshots/erofs/erofs.go

```diff
@@ -0,0 +1,622 @@
+		s              storage.Snapshot
+		td, path       string
+		info           snapshots.Info
+		containerUpper bool
+	)
+
+	var base = snapshots.Info{}
+	for _, opt := range opts {
+		if err := opt(&base); err != nil {
+			return nil, err
+		}
+	}
+	if _, ok := base.Labels[criContainerAnno]; ok {
+		containerUpper = true
+	}
+
+	defer func() {
+		if err != nil {
+			if td != "" {
+				if err1 := os.RemoveAll(td); err1 != nil {
+					log.G(ctx).WithError(err1).Warn("failed to cleanup temp snapshot directory")
+				}
+			}
+			if path != "" {
+				if err1 := os.RemoveAll(path); err1 != nil {
+					log.G(ctx).WithError(err1).WithField("path", path).Error("failed to reclaim snapshot directory, directory may need removal")
+					err = fmt.Errorf("failed to remove path: %v: %w", err1, err)
+				}
+			}
+		}
+	}()
+
+	if err := o.ms.WithTransaction(ctx, true, func(ctx context.Context) (err error) {
+		snapshotDir := filepath.Join(o.root, "snapshots")
+		td, err = o.prepareDirectory(ctx, snapshotDir, kind, containerUpper)
+		if err != nil {
+			return fmt.Errorf("failed to create prepare snapshot dir: %w", err)
+		}
+
+		s, err = storage.CreateSnapshot(ctx, kind, key, parent, opts...)
+		if err != nil {
+			return fmt.Errorf("failed to create snapshot: %w", err)
+		}
+
+		_, info, _, err = storage.GetInfo(ctx, key)
+		if err != nil {
+			return fmt.Errorf("failed to get snapshot info: %w", err)
+		}
+
+		path = filepath.Join(snapshotDir, s.ID)
+		if err = os.Rename(td, path); err != nil {
+			return fmt.Errorf("failed to rename: %w", err)
+		}
+		td = ""
+
+		if containerUpper {
+			if err := o.mergeDir(ctx, s, path); err != nil {
+				return fmt.Errorf("failed to do erofs merge: %w", err)
+			}
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	return o.mounts(ctx, s, info), nil
+}
+
+func (o *snapshotter) prepareDirectory(ctx context.Context, snapshotDir string, kind snapshots.Kind, containerUpper bool) (string, error) {
+	td, err := os.MkdirTemp(snapshotDir, "new-")
+	if err != nil {
+		return "", fmt.Errorf("failed to create temp dir: %w", err)
+	}
+
+	if containerUpper {
+		if err := os.Mkdir(filepath.Join(td, "fs"), 0755); err != nil {
+			return td, err
+		}
+
+		if kind == snapshots.KindActive {
+			if err := os.Mkdir(filepath.Join(td, "work"), 0711); err != nil {
+				return td, err
+			}
+		}
+	}
+
+	return td, nil
+}
+
+func (o *snapshotter) mounts(ctx context.Context, s storage.Snapshot, info snapshots.Info) []mount.Mount {
+	var options []string
+	merged := filepath.Join(o.plainPath(s.ID), "merged.erofs")
+
+	if _, err := os.Stat(merged); err != nil {
+		return []mount.Mount{
+			{
+				Source: o.plainPath(s.ID),
+				Type:   "bind",
+				Options: append(options,
+					"rbind",
+					"rw",
+				),
+			},
+		}
+	}
+
+	lowerDir := filepath.Join(o.plainPath(s.ID), "lower")
+	if _, err := os.Stat(lowerDir); os.IsNotExist(err) {
+		mergedLoop, err := o.loopDriver.Get(context.Background(), filepath.Join(o.plainPath(s.ID), "loop"))
```

Comment:
Why do you need context.Background, not ctx?

Reply:
originally that bc we did not want to corrupt original log in dev phase, fixed :)